src/Extension/Engine/SiteTreePublishingEngine.php

```diff
@@ -98,68 +99,100 @@ public function setToUpdate($toUpdate)
      */
-    public function onAfterPublishRecursive(&$original)
+    public function onBeforePublishRecursive($original)
     {
-        // If the site tree has been "reorganised" (ie: the parentID has changed)
-        // then this is the equivalent of an un-publish and publish as far as the
-        // static publisher is concerned
-        if ($original && (
-            // Intentionally using loose comparison as ParentID may contain either string or integer
-            $original->ParentID != $this->getOwner()->ParentID
-                || $original->URLSegment !== $this->getOwner()->URLSegment
-            )
+        // There is no original object. This might be the first time it has been published
+        if (!$original?->exists()) {
+            return;
+        }
+
+        // We want to find out if the URL for this page has changed at all. That can happen 2 ways: Either the page is
+        // moved in the SiteTree (ParentID changes), or the URLSegment is updated
+        // Apparently ParentID can sometimes be string, so make sure we cast to (int) for our comparison
+        if ((int) $original->ParentID !== (int) $this->getOwner()->ParentID
+            || $original->URLSegment !== $this->getOwner()->URLSegment
         ) {
+            // We have detected a change to the URL. We need to purge the old URLs for this page and any children
             $context = [
-                'action' => 'unpublish',
+                'action' => self::ACTION_UNPUBLISH,
             ];
             $original->collectChanges($context);
+            // We need to flushChanges() immediately so that Jobs are queued for the Pages while they still have their
+            // old URLs
             $original->flushChanges();
         }
+    }
+
+    /**
+     * @param SiteTree|SiteTreePublishingEngine|null $original
+     */
+    public function onAfterPublishRecursive($original)
+    {
+        $parentId = $original->ParentID ?? null;
+        $urlSegment = $original->URLSegment ?? null;
+
+        // Apparently ParentID can sometimes be string, so make sure we cast to (int) for our comparison
+        $parentChanged = $parentId && (int) $parentId !== (int) $this->getOwner()->ParentID;
+        $urlChanged = $urlSegment && $original->URLSegment !== $this->getOwner()->URLSegment;
+
         $context = [
-            'action' => 'publish',
+            'action' => self::ACTION_PUBLISH,
+            // If a URL change has been detected, then we need to force the recursive regeneration of all child
+            // pages
+            'urlChanged' => $parentChanged || $urlChanged,
         ];
+
         $this->collectChanges($context);
         $this->flushChanges();
     }
 
     public function onBeforeUnpublish()
     {
         $context = [
-            'action' => 'unpublish',
+            'action' => self::ACTION_UNPUBLISH,
         ];
         $this->collectChanges($context);
-    }
-
-    public function onAfterUnpublish()
```

Comment:
The reason this was needed is because previously this was not correctly collating URLs as they existed prior to the change occurring. It meant that you ended up with purge jobs with the **new** URL instead of the old URL.

I hadn't really considered the error state. Fair point. I'll have a go at resolving that.